ntoskrnl/se/token.c

```diff
@@ -135,34 +135,222 @@ SepDeleteTokenLock(
+ * Compares tokens if they're equal given their properties.
+ *
+ * @param[in] FirstToken
+ * The first token.
+ *
+ * @param[in] SecondToken
+ * The second token.
+ *
+ * @param[out] Equal
+ * The retrieved value which determines if the tokens are
+ * equal or not.
+ *
+ * @return
+ * Returns STATUS_SUCCESS.
+ */
+static
+NTSTATUS
+SepCompareTokens(
+    _In_ PTOKEN FirstToken,
+    _In_ PTOKEN SecondToken,
+    _Out_ PBOOLEAN Equal)
 {
     BOOLEAN Restricted, IsEqual = FALSE;
+    ULONG FirstTokenIndex, SecondTokenIndex;
+    PAGED_CODE();
 
     ASSERT(FirstToken != SecondToken);
 
     /* Lock the tokens */
     SepAcquireTokenLockShared(FirstToken);
     SepAcquireTokenLockShared(SecondToken);
 
-    /* FIXME: Check if every SID that is present in either token is also present in the other one */
+    /* Check if every SID that is present in either token is also present in the other one */
+    for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->UserAndGroupCount; FirstTokenIndex++)
+    {
+        for (SecondTokenIndex = 0; SecondTokenIndex < SecondToken->UserAndGroupCount; SecondTokenIndex++)
+        {
+            if (!SepCompareSidAndAttributesFromTokens(FirstToken,
+                                                      FirstTokenIndex,
+                                                      SecondToken,
+                                                      SecondTokenIndex))
+            {
+                DPRINT1("SepCompareTokens(): SIDs and/or attributes are not equal!\n");
+                goto Quit;
+            }
+        }
+    }
 
     Restricted = SeTokenIsRestricted(FirstToken);
     if (Restricted == SeTokenIsRestricted(SecondToken))
     {
         if (Restricted)
         {
-            /* FIXME: Check if every SID that is restricted in either token is also restricted in the other one */
+            /* Check if every SID that is restricted in either token is also restricted in the other one */
+            for (FirstTokenIndex = 0; FirstTokenIndex < FirstToken->RestrictedSidCount; FirstTokenIndex++)
+            {
+                for (SecondTokenIndex = 0; SecondTokenIndex < SecondToken->RestrictedSidCount; SecondTokenIndex++)
+                {
+                    if (!SepCompareRestrictedSid(FirstToken,
+                                                 FirstTokenIndex,
+                                                 SecondToken,
+                                                 SecondTokenIndex))
+                    {
+                        DPRINT1("SepCompareTokens(): Restricted SIDs are not equal!\n");
+                        goto Quit;
+                    }
+                }
+            }
         }
 
-        /* FIXME: Check if every privilege that is present in either token is also present in the other one */
-        DPRINT1("FIXME: Pretending tokens are equal!\n");
-        IsEqual = TRUE;
+        /* Check if every privilege that is present in either token is also present in the other one */
```

Comment:
Can you confirm that you want to look at the privileges and attributes from the tokens only if **just the first token** is restricted, but don't care about the second one? (That sounds broken.)